File: Student.java
import java.util.Date;

public class Student {
    private String studentId;
    private String emri;
    private String mbiemri;
    private String email;
    private String fjalëkalimi;
    private String roli = "Student";
    private String gjendjaLlogarise; // Aktive, Locked, UnderReview
    private int numTentativave;
    private Date datëRegjistrimi;
    private String departamenti;

    public Student() {}

    public Student(String studentId, String emri, String mbiemri,
                   String email, String fjalëkalimi, String gjendjaLlogarise,
                   String departamenti) {
        this.studentId       = studentId;
        this.emri            = emri;
        this.mbiemri         = mbiemri;
        this.email           = email;
        this.fjalëkalimi     = fjalëkalimi;
        this.gjendjaLlogarise = gjendjaLlogarise;
        this.departamenti    = departamenti;
        this.numTentativave  = 0;
        this.datëRegjistrimi = new Date();
    }

    public boolean login(String email, String pass) {
        return this.email.equals(email) && this.fjalëkalimi.equals(pass);
    }

    public void logout()              { System.out.println("Student " + emri + " doli."); }
    public boolean verifiko2FA(String kod) { return kod != null && kod.length() == 6; }
    public void resetFjalëkalim()     { System.out.println("Reset kërkuar për " + email); }
    public void rritTentativa()       { this.numTentativave++; }
    public void resetTentativa()      { this.numTentativave = 0; }

    // ── Getters & Setters ──────────────────────────────────────────────────
    public String getStudentId()         { return studentId; }
    public void   setStudentId(String v)  { this.studentId = v; }
    public String getEmri()              { return emri; }
    public void   setEmri(String v)       { this.emri = v; }
    public String getMbiemri()           { return mbiemri; }
    public void   setMbiemri(String v)    { this.mbiemri = v; }
    public String getEmail()             { return email; }
    public void   setEmail(String v)      { this.email = v; }
    public String getFjalëkalimi()       { return fjalëkalimi; }
    public void   setFjalëkalimi(String v){ this.fjalëkalimi = v; }
    public String getRoli()              { return roli; }
    public String getGjendjaLlogarise() { return gjendjaLlogarise; }
    public void   setGjendjaLlogarise(String v){ this.gjendjaLlogarise = v; }
    public int    getNumTentativave()    { return numTentativave; }
    public void   setNumTentativave(int v){ this.numTentativave = v; }
    public Date   getDatëRegjistrimi()   { return datëRegjistrimi; }
    public String getDepartamenti()      { return departamenti; }
    public void   setDepartamenti(String v){ this.departamenti = v; }

    @Override
    public String toString() {
        return "[" + studentId + "] " + emri + " " + mbiemri + " <" + email + "> — " + gjendjaLlogarise;
    }
}

File: python/university_security/models/student.py
"""Student domain class — mirrors Student.java."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

from .actor import Actor


@dataclass
class Student(Actor):
    """A university student account.

    Attributes
    ----------
    student_id:        Unique identifier (e.g. ``"S001"``).
    emri:              First name.
    mbiemri:           Last name.
    email:             University e-mail address.
    fjalekalimi:       Hashed / plain password (demo only).
    gjendja_llogarise: Account state — ``"Aktive"``, ``"Locked"``,
                       or ``"UnderReview"``.
    departamenti:      Academic department.
    num_tentativave:   Failed login attempt counter (default ``0``).
    date_regjistrimi:  Registration timestamp (default: *now*).
    """

    student_id: str
    emri: str
    mbiemri: str
    email: str
    fjalekalimi: str
    gjendja_llogarise: str
    departamenti: str
    num_tentativave: int = 0
    date_regjistrimi: datetime = field(default_factory=datetime.now)

    # ── Actor abstract property ───────────────────────────────────────────────
    @property
    def roli(self) -> str:
        return "Student"

    # ── Auth behaviours ───────────────────────────────────────────────────────
    def login(self, email: str, password: str) -> bool:
        """Return ``True`` when *email* and *password* match."""
        return self.email == email and self.fjalekalimi == password

    def logout(self) -> None:
        print(f"Student {self.emri} doli.")

    def verifiko_2fa(self, kod: str) -> bool:
        """Accept any non-``None`` six-character code."""
        return kod is not None and len(kod) == 6

    # ── Domain behaviours ─────────────────────────────────────────────────────
    def reset_fjalekalim(self) -> None:
        print(f"Reset kërkuar për {self.email}")

    def rrit_tentativa(self) -> None:
        """Increment the failed-login counter."""
        self.num_tentativave += 1

    def reset_tentativa(self) -> None:
        """Reset the failed-login counter to zero."""
        self.num_tentativave = 0

    # ── Dunder methods ────────────────────────────────────────────────────────
    def __str__(self) -> str:
        return (
            f"[{self.student_id}] {self.emri} {self.mbiemri}"
            f" <{self.email}> — {self.gjendja_llogarise}"
        )

    def __repr__(self) -> str:
        return (
            f"Student(student_id={self.student_id!r}, emri={self.emri!r},"
            f" mbiemri={self.mbiemri!r}, email={self.email!r},"
            f" gjendja_llogarise={self.gjendja_llogarise!r},"
            f" departamenti={self.departamenti!r})"
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Student):
            return NotImplemented
        return self.student_id == other.student_id

    def __hash__(self) -> int:
        return hash(self.student_id)
